src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
+    CheckReturn fin (a :--> b) = CheckReturn fin b
+    CheckReturn a b =
+        ( a ~ b
+        , TypeError
+            ( 'Text "Return type does not match:"
+                ':$$: 'Text "\tExpected \""
+                ':<>: 'ShowType a
+                ':<>: 'Text "\" but given function returns \""
+                ':<>: 'ShowType b
+                ':<>: 'Text "\""
+            )
+        )
+
+type family IsFinal (fin :: S -> Type) (p :: S -> Type) :: Bool where
+    IsFinal a a = 'True
+    IsFinal _ _ = 'False
+
+{- | Brings Plutarch function into Haskell given the final return type.
+     It will wrap each arguments and the return type with `TestableTerm`.
+
+ @since 2.1.0
 -}
-class FromPFunN (a :: S -> Type) (b :: S -> Type) (c :: PTermType) where
-    fromPFun' ::
-        Proxy c ->
-        (forall s. Term s (a :--> b)) ->
-        TestableFun c (a :--> b)
+class PUnLam' (fin :: S -> Type) (p :: S -> Type) (end :: Bool) where
+    pUnLam' :: (forall s. Term s p) -> PUnLamHask fin p
 
--- | @since 2.0.0
 instance
-    forall (a :: S -> Type) (b :: S -> Type).
-    ( PLift a
-    , PLift b
+    forall (fin :: S -> Type) (pa :: S -> Type) (pb :: S -> Type).
+    ( PUnLam' fin pb (IsFinal fin pb)
+    , PUnLamHask fin (pa :--> pb) ~ (TestableTerm pa -> PUnLamHask fin pb)
     ) =>
-    FromPFunN (a :--> b) PBool 'LastPFunction
+    PUnLam' fin (pa :--> pb) 'False
     where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) = TestableTerm $ f # x
+    pUnLam' f = \(TestableTerm y) -> pUnLam' @fin @pb @(IsFinal fin pb) (f # y)
 
--- | @since 2.0.0
-instance FromPFunN a PBool 'LastPTerm where
-    fromPFun' _ f (TestableTerm x) = TestableTerm $ f # x
+instance PUnLam' fin fin 'True where
+    pUnLam' x = TestableTerm x
 
--- | @since 2.0.0
-instance
-    forall
-        (a :: S -> Type)
-        (b :: S -> Type)
-        (c :: S -> Type)
-        (d :: S -> Type)
-        (e :: S -> Type).
-    ( c ~ (d :--> e)
-    , PLift a
-    , PLift b
-    , FromPFunN d e (PTT c)
-    ) =>
-    FromPFunN (a :--> b) c 'PFunction
-    where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) =
-        fromPFun' (Proxy @(PTT c)) $ f # x
+{- | Constraint for `PUnLam'` that will give a better type error message.
 
--- | @since 2.0.0
-instance
-    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type) (d :: S -> Type).
-    ( b ~ (c :--> d)
-    , FromPFunN c d (PTT b)
-    ) =>
-    FromPFunN a b 'PTerm
-    where
-    fromPFun' _ f (TestableTerm x) = fromPFun' (Proxy @(PTT b)) $ f # x
+ @since 2.1.0
+-}
+type PUnLam fin p = (PUnLam' fin p (IsFinal fin p), CheckReturn fin p)
+
+{- | Wrapper for @pUnLam'@. Only required the final return type.
+
+ @since 2.1.0
+-}
+punlam' ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam' = pUnLam' @fin @p @(IsFinal fin p)
+
+{- | Wrapper for @pUnLam'@. Same as @punlam'@ but evaluates the given
```

Comment:
is `pUnLam` exposed? If not, no need to reference it.

Reply:
I'm not sure if we should or should not expose these typeclasses.

Reply:
I would prefer not to if at all possible. I see them as implementation details, and they're _completely_ lawless.

Reply:
Okay that sounds good. We only need to expose constraint synonym for these--`PUnLam` and `PWrapLam`